### PR TITLE
[OSC-1236] - Minor useful logic changes and yak shaves

### PR DESCRIPTION
**Pull request by ChintanRaval**
logic changes:
- isolate logic that determines reasons for a model's full refresh
- add unit tests with CI builds
- add idempotency support between unit test runs

yak shaves:
- use constants for status messages
- use constants for audit column names
- use constants for sql table alises instead of inline strings
- move unused function to a static util method
- group meaningful / similar code together
- rename `data_source` to `source_db` and `target_engine` to `target_db`
- turn `rdl` into a proper module/package
- reformat non-`*.py` files
- autopep8 all the things

### CODE REVIEW

File: modules/DataLoadManager.py [hunk 1]
```diff
@@ -157,3 +149,28 @@ def start_single_import(self, target_engine, model_file, requested_full_refresh)
         data_load_tracker.completed_successfully()
         self.data_load_tracker_repository.save(data_load_tracker)
         self.logger.info(f"Import Complete for: {model_name}. {data_load_tracker.get_statistics()}")
+
+    @staticmethod
```
Comment (waltaro) [inline modules/DataLoadManager.py]: Is there a reason `*` is a parameter?

File: modules/Shared.py [hunk 2]
```diff
@@ -0,0 +1,18 @@
+class Constants:
+    APP_NAME = 'relational-data-loader'
```
Comment (waltaro) [inline modules/Shared.py]: Beautiful ❤️

File: modules/data_sources/MsSqlDataSource.py [hunk 3]
```diff
@@ -104,62 +104,61 @@ def get_next_data_frame(self, table_configuration, columns, batch_configuration,
...
-        self.logger.debug(f"Enabling ChangeTracking for {table_configuration['schema']}.{table_configuration['name']}\n"
-                          f"{sql_builder.getvalue()}")
-
-        self.database_engine.execute(text(sql_builder.getvalue()).execution_options(autocommit=True))
+        init_change_tracking_sql = "IF NOT EXISTS(SELECT 1 FROM sys.change_tracking_tables " \
+            f"WHERE object_id = OBJECT_ID('{table_configuration['schema']}.{table_configuration['name']}'))\n" \
+            "BEGIN\n" \
+            f"ALTER TABLE {table_configuration['schema']}.{table_configuration['name']} " \
+            f"ENABLE CHANGE_TRACKING WITH(TRACK_COLUMNS_UPDATED=OFF);\n" \
+            "END\n"
```
Comment (waltaro) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
            "END\n"

```

File: modules/DataLoadManager.py (hunk 1 above)
Comment (waltaro) [inline modules/DataLoadManager.py]: Is this method going to be used elsewhere in the future?

Comment (ChintanRaval) [reply modules/DataLoadManager.py]: Yes, some python magic. @seanbudd the magician.

File: modules/DataLoadManager.py [hunk 4]
```diff
@@ -42,76 +43,67 @@ def start_imports(self, target_engine, force_full_refresh_models):
...
 
-        if not full_refresh and change_tracking_info.force_full_load:
-            self.logger.info("Change tracking has forced this to be a full load")
-            full_refresh = True
-            full_refresh_reason = "Change Tracking Invalid"
+        last_successful_execution_exists = last_successful_data_load_execution is not None
+        model_changed = (not last_successful_execution_exists) or \
+                        (last_successful_data_load_execution.model_checksum != model_checksum)
+
+        full_refresh_reason, full_refresh = DataLoadManager.is_full_refresh(
```
[the remark this answers is not in the document]
Comment (ChintanRaval) [reply modules/DataLoadManager.py]: mmmmmm.. 😋

File: modules/DataLoadManager.py (hunk 1 above)
Comment (ChintanRaval) [reply modules/DataLoadManager.py]: nope

File: modules/DataLoadManager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (seanbudd) [reply modules/DataLoadManager.py]: hmm?

File: modules/DataLoadManager.py (hunk 1 above)
Comment (seanbudd) [reply modules/DataLoadManager.py]: ensures that you need to supply everything as keyword arguments (you *have* to call it like chintan did)

File: modules/DataLoadManager.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (ChintanRaval) [reply modules/DataLoadManager.py]: the function returns a tuple.

[the remark this answers is not in the document]
Comment (seanbudd) [reply modules/DataLoadManager.py]: mmmm

File: modules/DataLoadManager.py (hunk 1 above)
Comment (waltaro) [reply modules/DataLoadManager.py]: Do we need `@staticmethod`?

Comment (seanbudd) [reply modules/DataLoadManager.py]: 🤷‍♀️

File: modules/data_sources/MsSqlDataSource.py [hunk 5]
```diff
@@ -30,49 +32,46 @@ def prefix_column(column_name, full_refresh, primary_key_column_names):
...
-                              f" on {self.build_change_table_on_clause(batch_key_tracker)}")
-            sql_builder.write(f" WHERE {self.build_where_clause(batch_key_tracker, 'chg')}")
-            sql_builder.write(f" ORDER BY {order_by};")
-
-            return sql_builder.getvalue()
+            select_sql = f"SELECT TOP ({batch_config['size']}) {column_names}, " \
+                f"{MsSqlDataSource.CHANGE_TABLE_ALIAS}.SYS_CHANGE_VERSION AS data_pipeline_change_version, " \
+                f"CASE {MsSqlDataSource.CHANGE_TABLE_ALIAS}.SYS_CHANGE_OPERATION WHEN 'D' THEN 1 ELSE 0 " \
+                f"END AS data_pipeline_is_deleted \n"
+            from_sql = f" FROM CHANGETABLE(CHANGES" \
```
Comment (waltaro) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
            from_sql = f"FROM CHANGETABLE(CHANGES" \
```

File: modules/data_sources/MsSqlDataSource.py [hunk 6]
```diff
@@ -30,49 +32,46 @@ def prefix_column(column_name, full_refresh, primary_key_column_names):
...
+                f" {table_config['schema']}.{table_config['name']}," \
+                f" {change_tracking_info.this_sync_version})" \
+                f" AS {MsSqlDataSource.CHANGE_TABLE_ALIAS}" \
+                f" LEFT JOIN {table_config['schema']}.{table_config['name']} AS {MsSqlDataSource.SOURCE_TABLE_ALIAS}" \
+                f" ON {self.build_change_table_on_clause(batch_key_tracker)}"
```
Comment (waltaro) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
                f" ON {self.build_change_table_on_clause(batch_key_tracker)} "
```

File: modules/data_sources/MsSqlDataSource.py [hunk 7]
```diff
@@ -30,49 +32,46 @@ def prefix_column(column_name, full_refresh, primary_key_column_names):
...
+            where_sql = f" WHERE {self.build_where_clause(batch_key_tracker, MsSqlDataSource.CHANGE_TABLE_ALIAS)}"
```
Comment (waltaro) [inline modules/data_sources/MsSqlDataSource.py]:
```suggestion
            where_sql = f"WHERE {self.build_where_clause(batch_key_tracker, MsSqlDataSource.CHANGE_TABLE_ALIAS)}"
```

**Pull request merged**